### PR TITLE
GC+: Fix list objects from store

### LINKED ISSUE #4640: GC+: Use ParallelDataLister to read objects under `data/`
Data in lakeFS repositories is stored under the `data/` prefix. The gc listing doesn't list these files.

### LINKED ISSUE #4656: GC+: List root objects from namespace
Use NaiveDataLister and ingore the `data` folder
Add it to the removed objects calculation

### PR DESCRIPTION
**Pull request by N-o-Z**
Closes #4640
Closes #4656

### CODE REVIEW

File: clients/spark/core/src/main/scala/io/treeverse/gc/DataLister.scala [hunk 1]
```diff
@@ -60,7 +58,9 @@ class ParallelDataLister extends DataLister with Serializable {
       .map(_.path)
       .toDF("slice_id")
       .withColumn("udf", explode(objectsUDF(col("slice_id"))))
-      .withColumn("address", concat(col("slice_id"), lit("/"), col("udf._1")))
+      .withColumn("address",
```
Comment (johnnyaug) [inline clients/spark/core/src/main/scala/io/treeverse/gc/DataLister.scala:61]: Suggestion: simply make this a column of "basename", and construct the address column outside this function.

File: clients/spark/core/src/main/scala/io/treeverse/gc/UncommittedGarbageCollector.scala [hunk 2]
```diff
@@ -22,16 +22,32 @@ object UncommittedGarbageCollector {
...
+   *  @param storageNamespace The storageNamespace to read from
+   *  @param before Exclude objects which last_modified date is newer than before Date
+   *  @return DF listing all objects under given storageNamespace
+   */
   def listObjects(storageNamespace: String, before: Date): DataFrame = {
+    // TODO(niro): parallelize reads from root and data paths
     val sc = spark.sparkContext
-    val dataLocation = storageNamespace + "data" // TODO(niro): handle better
+    val dataPrefix = "data"
+    val dataLocation = storageNamespace + dataPrefix // TODO(niro): handle better
```
Comment (itaiad200) [inline clients/spark/core/src/main/scala/io/treeverse/gc/UncommittedGarbageCollector.scala:35]: Is the `TODO` still relevant?

File: clients/spark/core/src/main/scala/io/treeverse/gc/DataLister.scala [hunk 3]
```diff
@@ -60,7 +58,9 @@ class ParallelDataLister extends DataLister with Serializable {
...
+                  concat(lit(prefix), lit("/"), col("slice_id"), lit("/"), col("udf._1"))
```
Comment (itaiad200) [inline clients/spark/core/src/main/scala/io/treeverse/gc/DataLister.scala:62]: This is sorcery to me 🧹

File: clients/spark/core/src/main/scala/io/treeverse/gc/DataLister.scala [hunk 4]
```diff
@@ -11,21 +11,20 @@ import org.apache.hadoop.fs.FileStatus
...
-  override def listData(configMapper: ConfigMapper, path: Path): DataFrame = {
+  override def listData(configMapper: ConfigMapper, path: Path, prefix: String): DataFrame = {
     import spark.implicits._
     val fs = path.getFileSystem(configMapper.configuration)
     val dataIt = fs.listFiles(path, false)
     val dataList = new ListBuffer[(String, Long)]()
     while (dataIt.hasNext) {
       val fileStatus = dataIt.next()
-      dataList += ((fileStatus.getPath.toString, fileStatus.getModificationTime()))
+      dataList += ((fileStatus.getPath.getName, fileStatus.getModificationTime))
```
Comment (itaiad200) [inline clients/spark/core/src/main/scala/io/treeverse/gc/DataLister.scala:27]: What's the difference?

**Review by itaiad200 — APPROVED** [review]

File: clients/spark/core/src/main/scala/io/treeverse/gc/DataLister.scala (hunk 1 above)
Comment (N-o-Z) [reply clients/spark/core/src/main/scala/io/treeverse/gc/DataLister.scala:61]: Done

File: clients/spark/core/src/main/scala/io/treeverse/gc/UncommittedGarbageCollector.scala (hunk 2 above)
Comment (N-o-Z) [reply clients/spark/core/src/main/scala/io/treeverse/gc/UncommittedGarbageCollector.scala:35]: I want to keep it - maybe we need to pass the information from the server

File: clients/spark/core/src/main/scala/io/treeverse/gc/DataLister.scala (hunk 4 above)
Comment (N-o-Z) [reply clients/spark/core/src/main/scala/io/treeverse/gc/DataLister.scala:27]: toString provides the full url path, getName returns only the name

**Pull request merged**